### PR TITLE
feat: ability to list microvms based on name

### LINKED ISSUE #368: Ability to list microvms based on namespace and name
**Describe the solution you'd like:**
We are able to query a microvm spec without uid.

**Why do you want this feature:**

After we added UID (#291, #327) it's hard to get the microvm spec without knowing the generated UID. It's not an issue for automated processes, but gives a lot of pain mostly during debugging.

**Anything else you would like to add:**
The response should be a list of microvms as it is possible we have the same ns/name but with different uid, for example when cluster api re-creates resources, there is a window where the old and the new one still exists.

### LINKED ISSUE #370: List MicroVMs only returns one per namespace/name
Since we introduced the UID, mvm names are not unique. Which means we end up with this on disk:

```bash
$ ls /var/lib/flintlock/vm/ns0/mvm0
01FTBEN93VPVR44CKASKPBK7SB 
01FTBEP0Q58S98YGG1QN0XHSF9
01FTBENYT9A9CCY2GJ2E2G3RW2

$ ls /var/lib/flintlock/vm/ns0/mvm1/
01FTBEW8DGQXW8QTB5B3MX4E8T
01FTBF9CTP020V4PRM6HWSRTGW
```

This is not a problem in and of itself, but we need to update the List functionality so that it does not return the first thing under the `namespace/name` path.

Here is the output from a List after I have created 3 mvms called `mvm0` in `ns0`, and 2 called `mvm1` again in `ns0`. Only 2 are listed, one for each name.

<details>
<summary>List output</summary>

```bash
{
  "microvm": [
    {
      "version": 2,
      "spec": {
        "id": "mvm1",
        "namespace": "ns0",
        "vcpu": 2,
        "memory_in_mb": 2048,
        "kernel": {
          "image": "ghcr.io/weaveworks/flintlock-kernel:5.10.77",
          "filename": "boot/vmlinux",
          "add_network_config": true
        },
        "root_volume": {
          "id": "root",
          "mount_point": "/",
          "source": {
            "container_source": "ghcr.io/weaveworks/capmvm-kubernetes:1.21.8"
          },
          "partition_id": "",
          "size_in_mb": 0
        },
        "interfaces": [
          {
            "device_id": "mmds",
            "type": 1,
            "guest_mac": "AA:FF:00:00:00:01",
            "address": {
              "address": "169.254.0.1/16"
            }
          },
          {
            "device_id": "eth1",
            "guest_mac": ""
          }
        ],
        "metadata": {
          "meta-data": "aW5zdGFuY2VfaWQ6IG5zMC9tdm0xCmxvY2FsX2hvc3RuYW1lOiBtdm0xCnBsYXRmb3JtOiBsaXF1aWRfbWV0YWwKY2x1c3Rlcl9uYW1lOiAiIgo=",
          "user-data": "I2Nsb3VkLWNvbmZpZwpob3N0bmFtZTogbXZtMQp1c2VyczoKLSBuYW1lOiByb290CmZpbmFsX21lc3NhZ2U6IFRoZSBMaXF1aWQgTWV0YWwgYm9vdGVkIHN5c3RlbSBpcyBnb29kIHRvIGdvIGFmdGVyICRVUFRJTUUgc2Vjb25kcwpib290Y21kOgotIGxuIC1zZiAvcnVuL3N5c3RlbWQvcmVzb2x2ZS9zdHViLXJlc29sdi5jb25mIC9ldGMvcmVzb2x2LmNvbmYK"
        },
        "uid": "01FTBEW8DGQXW8QTB5B3MX4E8T"
      },
      "status": {
        "state": 1,
        "volumes": {
          "root": {
            "mount": {
              "source": "/dev/mapper/flintlock-thinpool-snap-125"
            }
          }
        },
        "kernel_mount": {
          "type": 1,
          "source": "/var/lib/containerd/io.containerd.snapshotter.v1.native/snapshots/112"
        },
        "network_interfaces": {
          "eth1": {
            "host_device_name": "flvtapfc667e2",
            "index": 227,
            "mac_address": "DA:41:07:07:E9:23"
          },
          "mmds": {
            "host_device_name": "fltap691240a",
            "index": 226,
            "mac_address": "F2:2B:B0:4D:84:40"
          }
        }
      }
    },
    {
      "version": 2,
      "spec": {
        "id": "mvm0",
        "namespace": "ns0",
        "vcpu": 2,
        "memory_in_mb": 2048,
        "kernel": {
          "image": "ghcr.io/weaveworks/flintlock-kernel:5.10.77",
          "filename": "boot/vmlinux",
          "add_network_config": true
        },
        "root_volume": {
          "id": "root",
          "mount_point": "/",
          "source": {
            "container_source": "ghcr.io/weaveworks/capmvm-kubernetes:1.21.8"
          },
          "partition_id": "",
          "size_in_mb": 0
        },
        "interfaces": [
          {
            "device_id": "mmds",
            "type": 1,
            "guest_mac": "AA:FF:00:00:00:01",
            "address": {
              "address": "169.254.0.1/16"
            }
          },
          {
            "device_id": "eth1",
            "guest_mac": ""
          }
        ],
        "metadata": {
          "meta-data": "aW5zdGFuY2VfaWQ6IG5zMC9tdm0wCmxvY2FsX2hvc3RuYW1lOiBtdm0wCnBsYXRmb3JtOiBsaXF1aWRfbWV0YWwKY2x1c3Rlcl9uYW1lOiAiIgo=",
          "user-data": "I2Nsb3VkLWNvbmZpZwpob3N0bmFtZTogbXZtMAp1c2VyczoKLSBuYW1lOiByb290CmZpbmFsX21lc3NhZ2U6IFRoZSBMaXF1aWQgTWV0YWwgYm9vdGVkIHN5c3RlbSBpcyBnb29kIHRvIGdvIGFmdGVyICRVUFRJTUUgc2Vjb25kcwpib290Y21kOgotIGxuIC1zZiAvcnVuL3N5c3RlbWQvcmVzb2x2ZS9zdHViLXJlc29sdi5jb25mIC9ldGMvcmVzb2x2LmNvbmYK"
        },
        "uid": "01FTBEN93VPVR44CKASKPBK7SB"
      },
      "status": {
        "state": 1,
        "volumes": {
          "root": {
            "mount": {
              "source": "/dev/mapper/flintlock-thinpool-snap-122"
            }
          }
        },
        "kernel_mount": {
          "type": 1,
          "source": "/var/lib/containerd/io.containerd.snapshotter.v1.native/snapshots/109"
        },
        "network_interfaces": {
          "eth1": {
            "host_device_name": "flvtap4944bf8",
            "index": 221,
            "mac_address": "1A:E6:30:3D:BA:50"
          },
          "mmds": {
            "host_device_name": "fltapd3a03ee",
            "index": 220,
            "mac_address": "E2:5B:49:DD:4C:A2"
          }
        }
      }
    }
  ]
}
```

</details>

### PR DESCRIPTION
**Pull request by yitsushi**
**What this PR does / why we need it**:

### Background

I considered two routes:

1. Create a new Find() method
2. Add the name field on the List() method

The second option sounds better because it's already a List method, so
makes sense to list mirovms that way.

The extra name field is optional, if it's not provided it works the same
way as before, but if it's provided it will filter based on namespace
and name.

#### proto and gRPC server

Changing the definition to use name in List queries.

#### tests

* Update all unit tests.
* Update e2e tests.
* Add extra case to e2e test: List request with name field and expect
  only one MicroVM in the response.

#### GetAll function chain

In the background I change the whole GetAll function chain to use a
filter `map[string]string`, that way we can use them to filter based on
different filters.

I considered doing a much cleaner solution something like this:

    // WithNamespace is a namespace filter for queries.
    func WithNamespace(namespace string) models.ListMicroVMQuery {
           return func() (string, string) {
                   return NamespaceLabel(), namespace
           }
    }

    // WithName is a name filter for queries.
    func WithName(name string) models.ListMicroVMQuery {
           return func() (string, string) {
                   return NameLabel(), namespace
           }
    }

But at the end, I couldn't find a good place for it becasue that should
live under containerd as those labels are containerd specific formats,
but then I couldn't think of an easy way to astract it through ports and
creating an outer layer for filters and put it in ports seems
unnecessary and overkill, so I stuck with the simple key-value map.

#### Fix list results

List endpoints returned only one of the MicroVMs when we have multiple MicroVMs with the same ns/name.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #368
Fixes #370

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by yitsushi** [comment]
```
--- PASS: TestE2E (30.16s)
PASS
ok      github.com/weaveworks/flintlock/test/e2e        30.162s
?       github.com/weaveworks/flintlock/test/e2e/utils  [no test files]
```

**Comment by yitsushi** [comment]
```
❯ cat hack/scripts/payload/ListMicroVMs.json \
  && echo "----" \
  && ./hack/scripts/send.sh -m ListMicroVMs \
    | dasel -p json --multiple '.microvm.[*].spec.id'
{
  "namespace": "ns1"
}
----
"mvm1"
"mvm2"

❯ cat hack/scripts/payload/ListMicroVMs.json \
  && echo "----" \
  && ./hack/scripts/send.sh -m ListMicroVMs \
    | dasel -p json --multiple '.microvm.[*].spec.id'
{
  "namespace": "ns1",
  "name": "mvm2"
}
----
"mvm2"
```

**Comment by yitsushi** [comment]
```
❯ cat hack/scripts/payload/ListMicroVMs.json \
  && echo "----" \
  && ./hack/scripts/send.sh -m ListMicroVMs \
    | dasel -p json --multiple '.microvm.[*].spec.uid'
{
  "namespace": "ns1",
  "name": "mvm2"
}
----
"01FTBC5ZEKE79FGYNRHPC8MME5"
"01FTBFG9KAC46TVDYGH1JMH665"

❯ cat hack/scripts/payload/ListMicroVMs.json \
  && echo "----" \
  && ./hack/scripts/send.sh -m ListMicroVMs \
    | dasel -p json --multiple '.microvm.[*].spec.uid'
{
  "namespace": "ns1"
}
----
"01FTBFG9KAC46TVDYGH1JMH665"
"01FTBC51BK5V82XEDBASXAF61Z"
"01FTBC5ZEKE79FGYNRHPC8MME5"
```

File: core/application/app_test.go [hunk 1]
```diff
@@ -409,11 +409,18 @@ func TestApp_GetAllMicroVM(t *testing.T) {
 		expect      func(rm *mock.MockMicroVMRepositoryMockRecorder, em *mock.MockEventServiceMockRecorder, im *mock.MockIDServiceMockRecorder, pm *mock.MockMicroVMServiceMockRecorder)
 	}{
 		{
-			name:        "empty namespace should return an error",
+			name:        "empty namespace should not return an error",
```
Comment (Callisto13) [inline core/application/app_test.go:412]: do we want equivalent test for name?

File: core/application/app_test.go [hunk 2]
```diff
@@ -454,7 +461,7 @@ func TestApp_GetAllMicroVM(t *testing.T) {
 			expect: func(rm *mock.MockMicroVMRepositoryMockRecorder, em *mock.MockEventServiceMockRecorder, im *mock.MockIDServiceMockRecorder, pm *mock.MockMicroVMServiceMockRecorder) {
 				rm.GetAll(
 					gomock.AssignableToTypeOf(context.Background()),
-					gomock.Eq("default"),
+					gomock.Eq(models.ListMicroVMQuery{"namespace": "default"}),
```
Comment (Callisto13) [inline core/application/app_test.go:464]: could not click on line 457 above, but do we want the name version of "microvms exist in namespace and are returned"

File: core/application/app_test.go (hunk 1 above)
Comment (yitsushi) [reply core/application/app_test.go:412]: A lot of them are testing that. And actually there is not test that receives the name.

**Review by Callisto13 — APPROVED** [review]

**Pull request merged**